src/kernels/webgl/reverse_packed_gpu.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {getChannels} from '../packing_util';
+import {GPGPUProgram} from './gpgpu_math';
+import {getCoordsDataType} from './shader_compiler';
+
+export class ReversePackedProgram implements GPGPUProgram {
+  variableNames = ['x'];
+  outputShape: number[];
+  userCode: string;
+  usesPackedTextures = true;
+
+  constructor(xShape: number[], axis: number[]) {
+    const rank = xShape.length;
+    if (rank > 4) {
+      throw new Error(
+          `WebGL backend: Reverse of rank-${rank} tensor is not yet supported`);
+    }
+    this.outputShape = xShape;
+    const channels = getChannels('rc', rank);
+    const nextColumn =
+        `++${channels[rank - 1]} < ${this.outputShape[rank - 1]}`;
+    const getInCoord = (i: number) => {
+      if (rank === 1) {
+        return `${xShape[0]} - rc - 1`;
+      } else if (axis.indexOf(i) !== -1 && xShape[i] !== 1) {
+        return `${xShape[i]} - rc[${i}] - 1`;
+      } else {
+        return `rc[${i}]`;
+      }
+    };
+    const inCoords = xShape.map((_, i) => getInCoord(i)).join(',');
+    const innerDims = rank === 1 ?
+        'rc' :
+        `vec2(${xShape.map((_, i) => getInCoord(i)).slice(-2).join(',')})`;
+    const type = getCoordsDataType(rank);
+    const getc = `getChannel(getX(${inCoords}), ${innerDims})`;
+
+    const upperRow = `
```

Comment:
What do you think about restructuring this shader to match the unpacked shader so that all the logic for rank = 1 is nested under one if-statement? I wonder whether it might be easier to read that way - avoids having to test for rank = 1 in so many places. If we did it this way we should also be able to inline `upperRow` and `lowerRow` in the body of `void main` below.

Reply:
Done